dev(watch): canonicalize project root & add artifact-aware ignores

navahas opened the pull request:
- Normalize project root once for stable path comparisons
- Build ignore set from project type + artifact parent dir
- Keep pause/reload semantics; filter before enqueue
- Add tests (rust/ts/mixed, artifact parent, rel vs abs, symlink, nested)

Closes #20

## Feature or Problem
Reduce self-triggered rebuilds in `wash dev` by ignoring generated outputs and comparing paths consistently. Canonicalizing the project root and adding project-aware ignore rules prevents noisy loops from `target/`, `dist/`, `node_modules/`, etc., and from the build artifact’s parent directory.

## Related Issues
- Closes #20

## Release Information
Included in the next release (no config or API changes).

## Consumer Impact
No breaking changes. Fewer spurious rebuilds; behavior is more predictable across platforms. Users may notice that edits inside build/output dirs no longer trigger rebuilds (intended).

## Testing

### Unit Test(s)
Added/updated tests to cover:
- Rust, TypeScript, Mixed contexts
- Artifact parent directory ignoring
- Relative vs absolute path consistency
- Symlink handling
- Nested ignored directories

### Acceptance or Integration
No changes required.

### Manual Verification
Using the patched binary, scaffold a tiny component and start the dev loop.

```bash
# Set your workspace dir (no spaces around =)
DIR="$HOME/dev/wasmCloud"

# Copy the patched wash binary next to your projects
cp "$DIR/wash/target/debug/wash" "$DIR/wash-fix"

# Scaffold a minimal Rust HTTP component (uses the local binary)
cd "$DIR"
./wash-fix new http-hello-world --template-name hello-world-rust
```
#### Terminal A
In the component folder
```
cd "$DIR/http-hello-world"
RUST_LOG=trace,wash=trace ../wash-fix dev --address 127.0.0.1:8000
```
#### Terminal B
same folder

Real source change
- Terminal A: SHOULD rebuild
```
echo "// change $(date)" >> src/lib.rs
```

Ignored dirs 
- Terminal A: SHOULD NOT rebuild
```
mkdir -p target/tmp dist/tmp node_modules/tmp .wash/tmp .wasmbuild/tmp
touch target/tmp/a.txt dist/tmp/b.txt node_modules/tmp/c.txt .wash/tmp/d.txt .wasmbuild/tmp/e.txt
```

Artifact parent dir
- Terminal A: SHOULD NOT rebuild
```
ART="$(ls target/wasm32-wasi*/debug/*.wasm | head -n1)"
touch "$ART" "$(dirname "$ART")/probe.txt"
```

Similar name, not ignored
- Terminal A: SHOULD rebuild
```
mkdir -p targetx
touch targetx/should-trigger.txt
```

navahas committed:
feat(dev/watch): canonicalize project root & add artifact-aware ignores

- Normalize project root once for stable path comparisons
- Build ignore set from project type + artifact parent dir
- Keep pause/reload semantics; filter before enqueue
- Add tests (rust/ts/mixed, artifact parent, rel vs abs, symlink, nested)

Closes #20

Signed-off-by: Carlos Navajas <c.navajas4@gmail.com>

navahas commented:
I used the tests as handy smoke checks while developing. I kept them in the PR in case others want to run them locally, feel free to remove or modify them.

brooksmtownsend commented:
Just a note @navahas im fixing the windows test issue in #36, so don't worry about that test specifically

brooksmtownsend commented on crates/wash/src/cli/dev.rs:
```diff
@@ -47,6 +47,140 @@ use crate::{
+            dirs_to_ignore.extend_from_slice(&["target", ".wasmbuild"]);
+        }
+        ProjectContext::TypeScript { .. } => {
+            dirs_to_ignore.extend_from_slice(&["node_modules", "dist", "build", ".next", ".nuxt"]);
```
Is `.nuxt` a typescript framework? For `.next`, does it ever include metadata that we should rebuild on?

navahas replied:
.nuxt is the build output from Nuxt (Vue-based, JS/TS). Similar to .next for Next.js, it only has generated code/metadata, so safe to ignore.

brooksmtownsend commented on crates/wash/src/cli/dev.rs:
```diff
@@ -47,6 +47,140 @@ use crate::{
+    // Add project-type-specific ignore patterns
+    match project_context {
+        ProjectContext::Rust { .. } => {
+            dirs_to_ignore.extend_from_slice(&["target", ".wasmbuild"]);
```
I think we can remove .wasmbuild here

navahas replied:
Ah, right! Forgot to remove on my end. It was an initial test placeholder for possible new build directories from future frameworks I might have missed in this PR. Safe to drop it.

brooksmtownsend replied:
Sounds good 😃

brooksmtownsend commented on crates/wash/src/cli/dev.rs:
```diff
@@ -47,6 +47,140 @@ use crate::{
+        }
+        ProjectContext::General => {
+            // For general context, include common patterns from all project types
+            dirs_to_ignore.extend_from_slice(&[
```
same comment re: removing .wasmbuild/.nuxt

brooksmtownsend commented on crates/wash/src/cli/dev.rs:
```diff
@@ -47,6 +47,140 @@ use crate::{
+                ignore_paths.insert(canonical);
+            }
+        }
+    }
```
Hrm, what happens if we're outputting the artifact directly into the source of the project?

brooksmtownsend replied:
I think that this is a good heuristic, but I think it may be best to leave it out and err on the side of an accidental rebuild once in a while

navahas replied:
Agree, an occasional extra rebuild is safer than missing one.

brooksmtownsend commented on crates/wash/src/cli/dev.rs:
```diff
@@ -47,6 +47,140 @@ use crate::{
+    }
+
+    // Common directories for all project types
+    let mut dirs_to_ignore = vec![".git", ".wash"];
```
I actually think if the config changes in `.wash`, we _do_ want to rebuild since it may affect the build

brooksmtownsend commented on crates/wash/src/cli/dev.rs:
```diff
@@ -540,3 +724,224 @@ pub async fn handle_request<'a>(
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
```
tests 😍

brooksmtownsend commented on crates/wash/src/cli/dev.rs:
```diff
@@ -250,14 +384,41 @@ impl CliCommand for DevCommand {
+        let ignore_paths = Arc::new(RwLock::new(initial_ignore_set));
+        let ignore_paths_notify = ignore_paths.clone();
+
+        let canonical_project_root_notify = canonical_project_root.clone();
```
I might be missing some rationale here, but I think that removing the RwLock and not worrying about inserting an artifact path later on would make this quite a bit simpler since I think the artifact path should be nested under a path we ignore already

brooksmtownsend commented on crates/wash/src/cli/dev.rs:
```diff
@@ -47,6 +47,140 @@ use crate::{
+    if let Ok(relative_path) = canonical_path.strip_prefix(canonical_project_root) {
+        // Check if any component of the relative path matches common build directories
+        for component in relative_path.components() {
+            if let Some(component_str) = component.as_os_str().to_str() {
```
I think I'd like to skip this loop / recursive checking for the path components since we are providing a set of ignored paths already to this function, and that would help eliminate some confusing edge cases

navahas replied:
Fair, recursion is something I struggle to agree with too, but here I wasn’t sure how else to reliably catch nested cases like subproject/target without listing every possible nested path in the ignore set. That was the main reason I added it, though I get it adds edge cases.

brooksmtownsend reviewed (changes requested):
I had a lot of comments for this one @navahas but this looks really nice! My requests are all around removing some of the logic that might ignore more than we want, I'd rather trigger a couple extra rebuilds than ignore a legitimate change.

If you ran into anything when developing this, please feel free to push back on my suggestions

navahas commented:
> I had a lot of comments for this one @navahas but this looks really nice! My requests are all around removing some of the logic that might ignore more than we want, I'd rather trigger a couple extra rebuilds than ignore a legitimate change.
> 
> If you ran into anything when developing this, please feel free to push back on my suggestions

Appreciate! Honestly, I’m just happy to contribute with an initial PR here, I’m still getting familiar with the whole project, and you’ve got a much better sense of the bigger picture. I’m happy for any ideas or modifications you think fit better.

navahas committed:
fix(watch): apply review feedback
- Canonicalize project root once.
- Compare paths with `starts_with` against ignore set.
- Ignore dirs from project type only; keep `.wash/` watched.
- Drop artifact-parent rule and RwLock.

Tests:
- Rust/TS/Mixed contexts.
- Rel vs abs; symlinks.
- Nested subproject `target/` not ignored.

Signed-off-by: Carlos Navajas <c.navajas4@gmail.com>

navahas commented:
## Manual Verification

### Terminal A
In the component folder
```wash dev .```

### Terminal B

#### Real source change
- Terminal A (wash dev): SHOULD rebuild
```bash
echo "// change $(date)" >> src/lib.rs
```

#### Ignored dirs
- Terminal A (wash dev): SHOULD NOT rebuild
```bash
mkdir -p target/tmp node_modules/tmp
touch target/tmp/as.txt node_modules/tmp/cs.txt
```

#### Relevant dirs
- Terminal A (wash dev): SHOULD rebuild
```bash
mkdir -p dist/tmp_2
touch dist/tmp_2/bs.txt
```
```bash
mkdir -p .wash/tmp_dir
touch .wash/tmp_dir/log.txt
```

#### Similar name, not ignored
- Terminal A: SHOULD rebuild
```bash
mkdir -p targetx
touch targetx/should-trigger.txt
```

brooksmtownsend reviewed (approved):
LGTM @navahas ! Looks like you just need a rebase

navahas commented:
sorted!
